libsel4vm/src/arch/arm/vgic/virq.h

```diff
@@ -0,0 +1,225 @@
+    /* Queue for IRQs that don't fit in the GIC's vCPU list registers */
+    struct irq_queue irq_queue;
+    /*  vCPU local interrupts (SGI, PPI) */
+    virq_handle_t local_virqs[NUM_VCPU_LOCAL_VIRQS];
+} vgic_vcpu_t;
+
+/* GIC global interrupt context */
+typedef struct vgic {
+    /* virtual distributor registers */
+    struct gic_dist_map *dist;
+    /* registered global interrupts (SPI) */
+    virq_handle_t vspis[NUM_SLOTS_SPI_VIRQ];
+    /* vCPU specific interrupt context */
+    vgic_vcpu_t vgic_vcpu[CONFIG_MAX_NUM_NODES];
+} vgic_t;
+
+static inline vgic_vcpu_t *get_vgic_vcpu(vgic_t *vgic, int vcpu_id)
+{
+    assert(vgic);
+    assert((vcpu_id >= 0) && (vcpu_id < ARRAY_SIZE(vgic->vgic_vcpu)));
+    return &(vgic->vgic_vcpu[vcpu_id]);
+}
+
+static inline struct virq_handle *virq_get_sgi_ppi(vgic_t *vgic, vm_vcpu_t *vcpu, int virq)
+{
+    vgic_vcpu_t *vgic_vcpu = get_vgic_vcpu(vgic, vcpu->vcpu_id);
+    assert(vgic_vcpu);
+    assert((virq >= 0) && (virq < ARRAY_SIZE(vgic_vcpu->local_virqs)));
+    return vgic_vcpu->local_virqs[virq];
+}
+
+static inline struct virq_handle *virq_find_spi_irq_data(struct vgic *vgic, int virq)
+{
+    for (int i = 0; i < ARRAY_SIZE(vgic->vspis); i++) {
+        if (vgic->vspis[i] && vgic->vspis[i]->virq == virq) {
+            return vgic->vspis[i];
+        }
+    }
+    return NULL;
+}
+
+static inline struct virq_handle *virq_find_irq_data(struct vgic *vgic, vm_vcpu_t *vcpu, int virq)
+{
+    if (virq < NUM_VCPU_LOCAL_VIRQS)  {
+        return virq_get_sgi_ppi(vgic, vcpu, virq);
+    }
+    return virq_find_spi_irq_data(vgic, virq);
+}
+
+static inline int virq_spi_add(vgic_t *vgic, struct virq_handle *virq_data)
+{
+    for (int i = 0; i < ARRAY_SIZE(vgic->vspis); i++) {
+        if (vgic->vspis[i] == NULL) {
+            vgic->vspis[i] = virq_data;
+            return 0;
+        }
+    }
+    return -1;
+}
+
+static inline int virq_sgi_ppi_add(vm_vcpu_t *vcpu, vgic_t *vgic, struct virq_handle *virq_data)
+{
+    vgic_vcpu_t *vgic_vcpu = get_vgic_vcpu(vgic, vcpu->vcpu_id);
+    assert(vgic_vcpu);
+    int irq = virq_data->virq;
+    assert((irq >= 0) && (irq < ARRAY_SIZE(vgic_vcpu->local_virqs)));
+    virq_handle_t *slot = &vgic_vcpu->local_virqs[irq];
+    if (*slot != NULL) {
+        ZF_LOGE("IRQ %d already registered on VCPU %u", virq_data->virq, vcpu->vcpu_id);
+        return -1;
+    }
+    *slot = virq_data;
+    return 0;
+}
+
+static inline int virq_add(vm_vcpu_t *vcpu, vgic_t *vgic, struct virq_handle *virq_data)
+{
+    int virq = virq_data->virq;
+    if (virq < NUM_VCPU_LOCAL_VIRQS) {
+        return virq_sgi_ppi_add(vcpu, vgic, virq_data);
+    }
+    return virq_spi_add(vgic, virq_data);
+}
+
+static inline void virq_init(virq_handle_t virq, int irq, irq_ack_fn_t ack_fn, void *token)
+{
+    virq->virq = irq;
+    virq->token = token;
+    virq->ack = ack_fn;
+}
+
+static inline struct vgic *vgic_device_get_vgic(struct vgic_dist_device *d)
```

Comment:
This function does not really belong here, it would fit better in `vgic_v2.c` even if this duplicates code for GICv3 then. I have a branch where `struct vgic_dist_device` is dropped completely and this is all part of the `struct vgic` then, but have to check how this would work with the GICv3 code then.

Reply:
What you say makes sense about the function staying in vgic_v2.c, I need to check with the old version whether it had it in there or whether I accidentally pulled it in during a rebase. 

I'm not sure whether dropping vgic_dist_device would make the code cleaner and it'd be better to wait until v3 code is merged before we consider IMO.

Reply:
I added a new commit to address this: https://github.com/seL4/seL4_projects_libs/pull/68/commits/0ec24c1c8c13a73a7441bd9290dbab5a46c71424.  I think it's a bit reactive, but still reasonable to fix it this way.